### ISSUE TITLE
storage: TestNodeLivenessSetDecommissioning fatals under stress

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#39660]
While experimenting with shortening the closed timestamp target_duration in #39643 I encountered failures while running `TestNodeLivenessSetDecomissioning` under stress or stressrace leads to a fatal event. The full logs are attached below. The failure readily reproduces after 1000s of runs. 
I've reproduced this failure as far back as 9c8e44e29c6aeec77564af32ccbb21b14f97ce57 so this is not new. It generally takes thousands of runs under stress to reproduce.

Off of 1a48f102d35ba8dabfb5b8eac7ad66b6bb9e4248:

```
I190814 13:16:32.434790 1641 storage/store_snapshot.go:786  [replicate,s1,r6/1:/Table/{SystemCon…-11}] sending LEARNER snapshot 0243daa0 at applied index 16                                                                                                                   
I190814 13:16:32.435061 1641 storage/store_snapshot.go:829  [replicate,s1,r6/1:/Table/{SystemCon…-11}] streamed snapshot to (n2,s2):2: kv pairs: 44, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s                                                                            
I190814 13:16:32.435720 2327 storage/replica_raftstorage.go:823  [s2,r6/2:{-}] applying LEARNER snapshot at index 16 (id=0243daa0, encoded size=8206, 1 rocksdb batches, 0 log entries)                                                                                        
I190814 13:16:32.436232 2327 storage/replica_raftstorage.go:829  [s2,r6/2:/Table/{SystemCon…-11}] applied LEARNER snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]                                                                                                 
I190814 13:16:32.437799 1641 storage/replica_command.go:1198  [replicate,s1,r6/1:/Table/{SystemCon…-11}] change replicas (ADD_REPLICA (n2,s2):2): existing descriptor r6:/Table/{SystemConfigSpan/Start-11} [(n1,s1):1, (n2,s2):2LEARNER, next=3, gen=1]  
```

Off of 9c8e44e29c6aeec77564af32ccbb21b14f97ce57: 
```
I190814 14:29:54.295686 69 gossip/gossip.go:396  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:32816" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
W190814 14:29:54.303664 1805 storage/store.go:1532  [s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
W190814 14:29:54.374987 69 gossip/gossip.go:1500  [n2] no incoming or outgoing connections
I190814 14:29:54.375132 69 gossip/gossip.go:396  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:46215" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I190814 14:29:54.375260 2393 gossip/client.go:126  [n2] started gossip client to 127.0.0.1:32816
W190814 14:29:54.489461 69 gossip/gossip.go:1500  [n3] no incoming or outgoing connections
I190814 14:29:54.489523 69 gossip/gossip.go:396  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:36787" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I190814 14:29:54.489717 2886 gossip/client.go:126  [n3] started gossip client to 127.0.0.1:32816
I190814 14:29:54.492072 69 storage/client_test.go:493  gossip network initialized
I190814 14:29:54.493488 2235 storage/store_snapshot.go:784  [replicate,s1,r6/1:/Table/{SystemCon…-11}] sending PREEMPTIVE snapshot b207e0e8 at applied index 13
I190814 14:29:54.493643 2235 storage/store_snapshot.go:827  [replicate,s1,r6/1:/Table/{SystemCon…-11}] streamed snapshot to (n2,s2):?: kv pairs: 43, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190814 14:29:54.494146 2736 storage/replica_raftstorage.go:810  [s2,r6/?:{-}] applying PREEMPTIVE snapshot at index 13 (id=b207e0e8, encoded size=7969, 1 rocksdb batches, 0 log entries)
I190814 14:29:54.494378 69 util/stop/stopper.go:544  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190814 14:29:54.494400 69 util/stop/stopper.go:544  quiescing; tasks left:
1      [async] intent_resolver_gc_batcher
I190814 14:29:54.518555 3241 storage/node_liveness.go:476  [liveness-hb] heartbeat failed on epoch increment; retrying
I190814 14:29:54.519511 2736 storage/replica_raftstorage.go:816  [s2,r6/?:/Table/{SystemCon…-11}] applied PREEMPTIVE snapshot in 25ms [clear=0ms batch=0ms entries=0ms commit=24ms]
I190814 14:29:54.520617 2235 storage/replica_command.go:882  [replicate,s1,r6/1:/Table/{SystemCon…-11}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r6:/Table/{SystemConfigSpan/Start-11} [(n1,s1):1, next=2, gen=0]
I190814 14:29:54.523179 2235 storage/replica_raft.go:397  [replicate,s1,r6/1:/Table/{SystemCon…-11},txn=d4bd9269] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I190814 14:29:54.524870 2235 storage/store_snapshot.go:784  [replicate,s1,r6/1:/Table/{SystemCon…-11}] sending PREEMPTIVE snapshot 9f6cfeed at applied index 15
I190814 14:29:54.525063 2235 storage/store_snapshot.go:827  [replicate,s1,r6/1:/Table/{SystemCon…-11}] streamed snapshot to (n3,s3):?: kv pairs: 45, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190814 14:29:54.526385 2792 storage/replica_raftstorage.go:810  [s3,r6/?:{-}] applying PREEMPTIVE snapshot at index 15 (id=9f6cfeed, encoded size=8151, 1 rocksdb batches, 0 log entries)
I190814 14:29:54.527153 2792 storage/replica_raftstorage.go:816  [s3,r6/?:/Table/{SystemCon…-11}] applied PREEMPTIVE snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=1ms]
I190814 14:29:54.528129 2235 storage/replica_command.go:882  [replicate,s1,r6/1:/Table/{SystemCon…-11}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r6:/Table/{SystemConfigSpan/Start-11} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
F190814 14:29:54.529882 3204 storage/replica_raft.go:698  [s2,r6/2:{-}] setting non-zero HardState.Commit on uninitialized replica [n2,s2,r6/2:{-}]. HS={Term:6 Vote:0 Commit:15 XXX_unrecognized:[]}
```

[TestNodeLivenessSetDecommissioning-logs.txt](https://github.com/cockroachdb/cockroach/files/3501594/TestNodeLivenessSetDecommissioning-logs.txt)

**Comment by ajwerner** [comment]
`make roachprod-stress PKG=./pkg/storage TESTS=TestNodeLivenessSetDecommissioning CLUSTER=...`

The above command has been my tool to reproduce. I created a cluster of 4 16-core machines. 

Here's the current state of the `git bisect log`

```
git bisect start
# bad: [5fb59c4a1baeb09d0cb1e08613d0c4598fce1647] Merge #39657
git bisect bad 5fb59c4a1baeb09d0cb1e08613d0c4598fce1647
# bad: [9c8e44e29c6aeec77564af32ccbb21b14f97ce57] roachtest: remove tests from hibernate blacklist
git bisect bad 9c8e44e29c6aeec77564af32ccbb21b14f97ce57
# bad: [292972aef44d433a05beb85ac03219820916f5b1] roachtest: fix `fixtures import` on release branches
git bisect bad 292972aef44d433a05beb85ac03219820916f5b1
# good: [04dbe335735f8879851c6819a76e517ec00ad3b3] Merge #34890
git bisect good 04dbe335735f8879851c6819a76e517ec00ad3b3
# good: [346b134ac575002b9aeb07bbf64a352aa1f78dd9] Merge #36343
git bisect good 346b134ac575002b9aeb07bbf64a352aa1f78dd9
# bad: [43c3cf31832ec6502700f5b0a8955988154061ef] Merge #37011
git bisect bad 43c3cf31832ec6502700f5b0a8955988154061ef
```

In order to get this to continue to build, sometimes IGNORE_GOVERS=1 is required and some judicious removal of build artifacts like `.eg.go` files.

**Comment by ajwerner** [comment]
Bisect tracks this flake down to:
[1cd61f9fa27ebcd7c16d79790ea4083b8c8f3666] storage: use async methods for queuing ranges 

The second commit from https://github.com/cockroachdb/cockroach/pull/36564.

Here's the full bisect log:

```git bisect start
# bad: [5fb59c4a1baeb09d0cb1e08613d0c4598fce1647] Merge #39657
git bisect bad 5fb59c4a1baeb09d0cb1e08613d0c4598fce1647
# bad: [9c8e44e29c6aeec77564af32ccbb21b14f97ce57] roachtest: remove tests from hibernate blacklist
git bisect bad 9c8e44e29c6aeec77564af32ccbb21b14f97ce57
# bad: [292972aef44d433a05beb85ac03219820916f5b1] roachtest: fix `fixtures import` on release branches
git bisect bad 292972aef44d433a05beb85ac03219820916f5b1
# good: [04dbe335735f8879851c6819a76e517ec00ad3b3] Merge #34890
git bisect good 04dbe335735f8879851c6819a76e517ec00ad3b3
# good: [346b134ac575002b9aeb07bbf64a352aa1f78dd9] Merge #36343
git bisect good 346b134ac575002b9aeb07bbf64a352aa1f78dd9
# bad: [43c3cf31832ec6502700f5b0a8955988154061ef] Merge #37011
git bisect bad 43c3cf31832ec6502700f5b0a8955988154061ef
# bad: [d2317c962395f3e52671918da88fb3fdc35bd4dd] Merge #36693
git bisect bad d2317c962395f3e52671918da88fb3fdc35bd4dd
# good: [762a6a54bac116085326b817b0d5ed7bc0631ec5] Merge #36445 #36486 #36502
git bisect good 762a6a54bac116085326b817b0d5ed7bc0631ec5
# bad: [5c2cc21dbf84cce2f96f56a51337064c5474d163] Merge #36612
git bisect bad 5c2cc21dbf84cce2f96f56a51337064c5474d163
# good: [71d2c4cd98e16e4f8d684588ccbb7b7d115a2446] Merge #36509
git bisect good 71d2c4cd98e16e4f8d684588ccbb7b7d115a2446
# good: [d6a33f19b242d800a80ffa4b1af2543919083add] Merge #36571
git bisect good d6a33f19b242d800a80ffa4b1af2543919083add
# bad: [f8be5091ee4324542d18b2f6bb93ef9e6ec139eb] Merge #36564
git bisect bad f8be5091ee4324542d18b2f6bb93ef9e6ec139eb
# good: [58c458efeaa3b38c8c982f23a36381aac1b1004b] Merge #36565
git bisect good 58c458efeaa3b38c8c982f23a36381aac1b1004b
# good: [4eacf37155b184f0fe5cd0b6abb638f67932aac3] Merge #36595
git bisect good 4eacf37155b184f0fe5cd0b6abb638f67932aac3
# bad: [1cd61f9fa27ebcd7c16d79790ea4083b8c8f3666] storage: use async methods for queuing ranges
git bisect bad 1cd61f9fa27ebcd7c16d79790ea4083b8c8f3666
# good: [2a94c46a7c6a0e425544869f53ce26d7665eee2c] storage: use fine-grained locks in baseQueue.mu
git bisect good 2a94c46a7c6a0e425544869f53ce26d7665eee2c
# first bad commit: [1cd61f9fa27ebcd7c16d79790ea4083b8c8f3666] storage: use async methods for queuing ranges

```

**Comment by ajwerner** [comment]
Anecdotally I'm having a hard time reproducing the flake at the first reported commit on my gceworker using `make stress` but an easy time with `make roachprod-stress`. Nevertheless it reproduces locally on master. Maybe it has become more common? I'm going to try another round of bisect with only local stress given how quickly it seems to happen on master to see where it became more likely.

**Comment by ajwerner** [comment]
That was a less than satisfying experiment. Bisecting with a 3m local timeout led to:

```
# bad: [5fb59c4a1baeb09d0cb1e08613d0c4598fce1647] Merge #39657
git bisect bad 5fb59c4a1baeb09d0cb1e08613d0c4598fce1647
# good: [1cd61f9fa27ebcd7c16d79790ea4083b8c8f3666] storage: use async methods for queuing ranges
git bisect good 1cd61f9fa27ebcd7c16d79790ea4083b8c8f3666
# good: [6f7113046584a432eeaef9080e45a278a6934990] Merge #38014
git bisect good 6f7113046584a432eeaef9080e45a278a6934990
# bad: [6092913c0f8d8d8ccba68f1a649f0d80e0d5374c] Merge #38938
git bisect bad 6092913c0f8d8d8ccba68f1a649f0d80e0d5374c
# bad: [15d4329e3e38900790c2139ebefe0cbc97c862ad] Merge #38530
git bisect bad 15d4329e3e38900790c2139ebefe0cbc97c862ad
# good: [27fe883b6205471e4b8a2f6c97fad12d2cee2cd2] Merge #38400
git bisect good 27fe883b6205471e4b8a2f6c97fad12d2cee2cd2
# bad: [97b839533343ca3a00c642c6eae7196e5eb9944e] Merge #38465
git bisect bad 97b839533343ca3a00c642c6eae7196e5eb9944e
# bad: [ace062479d9b745b8429acb6213bd6dd56069c68] Merge #38419
git bisect bad ace062479d9b745b8429acb6213bd6dd56069c68
# bad: [cb15ab6a3f1994cc7c6cc106ec56caf96f8a1492] Merge #38421
git bisect bad cb15ab6a3f1994cc7c6cc106ec56caf96f8a1492
# bad: [b381e11993cb35d293de507c450e3cc6445fb718] Merge #38382
git bisect bad b381e11993cb35d293de507c450e3cc6445fb718
# bad: [a2b3d7050c934c7d2fe561d39da04d196e14565f] Merge #38406
git bisect bad a2b3d7050c934c7d2fe561d39da04d196e14565f
# bad: [197d1cd3c7b151db65a1d41e9c0dc41b16b6adc2] Merge #38216
git bisect bad 197d1cd3c7b151db65a1d41e9c0dc41b16b6adc2
# bad: [cb7b5505f3d4bdf595799d0d894887b0bf1c38ab] sql: Implement table reader that performs a loose index scan
git bisect bad cb7b5505f3d4bdf595799d0d894887b0bf1c38ab
# first bad commit: [cb7b5505f3d4bdf595799d0d894887b0bf1c38ab] sql: Implement table reader that performs a loose index scan
```

Which makes no sense. I can try again with a 10m timeout?

**Comment by ajwerner** [comment]
Running with a longer timeout again revealed https://github.com/cockroachdb/cockroach/pull/36564 to be the culprit. I'm going to leave this now.

**Comment by irfansharif** [comment]
The failure seen here was the same as seen in #40213, which was fixed with #40899. As a sanity check I tried stressing this and saw nothing.
